ink/contracts/test_oracle/lib.rs

```diff
@@ -0,0 +1,963 @@
+            // set the owner of this contract
+            ownable::Internal::_init_with_owner(&mut instance, caller);
+            // set the admin of this contract
+            access_control::Internal::_init_with_admin(&mut instance, Some(caller));
+            // grant the role manager
+            AccessControl::grant_role(&mut instance, MANAGER_ROLE, Some(caller))
+                .expect("Should grant the role MANAGER_ROLE");
+            instance
+        }
+
+        #[ink(message)]
+        #[openbrush::modifiers(access_control::only_role(MANAGER_ROLE))]
+        pub fn create_trading_pair(
+            &mut self,
+            trading_pair_id: TradingPairId,
+            token0: String,
+            token1: String,
+        ) -> Result<(), ContractError> {
+            // we create a new trading pair or override an existing one
+            let trading_pair = TradingPair {
+                token0,
+                token1,
+                value: 0,
+                nb_updates: 0,
+                last_update: 0,
+            };
+            self.trading_pairs.insert(trading_pair_id, &trading_pair);
+            Ok(())
+        }
+
+        #[ink(message)]
+        #[openbrush::modifiers(access_control::only_role(MANAGER_ROLE))]
+        pub fn request_price(
+            &mut self,
+            trading_pair_id: TradingPairId,
+        ) -> Result<QueueIndex, ContractError> {
+            let index = match self.trading_pairs.get(trading_pair_id) {
+                Some(t) => {
+                    // push the message in the queue
+                    let message = PriceRequestMessage {
+                        trading_pair_id,
+                        token0: t.token0,
+                        token1: t.token1,
+                    };
+                    self.push_message(&message)?
+                }
+                _ => return Err(ContractError::MissingTradingPair),
+            };
+
+            Ok(index)
+        }
+
+        #[ink(message)]
+        pub fn get_trading_pair(&self, trading_pair_id: TradingPairId) -> Option<TradingPair> {
+            self.trading_pairs.get(trading_pair_id)
+        }
+
+        #[ink(message)]
+        pub fn register_attestor(
+            &mut self,
+            account_id: AccountId,
+            ecdsa_public_key: [u8; 33],
+        ) -> Result<(), ContractError> {
+            AccessControl::grant_role(self, ATTESTOR_ROLE, Some(account_id))?;
+            self.register_ecdsa_public_key(account_id, ecdsa_public_key)?;
+            Ok(())
+        }
+
+        #[ink(message)]
+        pub fn get_attestor_role(&self) -> RoleType {
+            ATTESTOR_ROLE
+        }
+
+        #[ink(message)]
+        pub fn get_manager_role(&self) -> RoleType {
+            MANAGER_ROLE
+        }
+    }
+
+    impl KvStore for TestOracle {}
+    impl MessageQueue for TestOracle {}
+    impl RollupAnchor for TestOracle {}
+    impl MetaTransaction for TestOracle {}
```

Comment:
I'm very happy that you have considered the granularity here. I also started with a plain kv store first, and then an inherited constract to add queue functionalities. However, after two iterations, I think it's unlikely that the rollup anchor, the kvstore, and the queue would be separate, given all those features are based on the versioned kvstore. So my suggestion is to combine them together.

MetaTx is different. Someone may never use its functionalities. I still combined them together in the EVM version because it only increase the code size a little bit, not too much. So generally it simplify everything.

Reply:
I understand your point of view. 
Can we plan these changes in the milestone 3?